### PR TITLE
[Review/DontMergeYet] All sorts of things

### PR DESCRIPTION
**Pull request by damiel**
This pull request ports over alot of graphical features from rbdoom3bfg including HDR, SRGB, linear HDR, SMAA, SSAO, SSGI, Tonemaps and a bunch of fixes and ReadMe updates.

 I know there is still a bunch left from rbdoom3bfg, but i cant port those over yet as i dont uderstand those all too well, i tried porting them over but faced lots of problems so i put those on hold for now.

I think its now a good time for me to work on renderfixes rather then adding more new stuff :)

### CODE REVIEW

File: neo/d3xp/menus/MenuScreen.h [hunk 1]
```diff
@@ -33,6 +33,1830 @@ If you have questions concerning this license or the applicable additional terms
 #include "../cegui/MenuScreen_CEGUI.h"
 #else
 #include "../d3xp/menus/MenuScreen_SWF.h"
+// damiel: Not sure if we want that here
```
Comment (kortemik) [inline neo/d3xp/menus/MenuScreen.h]: I think we have this content in the _SWF.h file for flash implementation

File: neo/renderer/RenderLog.cpp [hunk 2]
```diff
@@ -528,7 +529,7 @@ void idRenderLog::LogCloseBlock( renderLogIndentLabel_t label )
 {
 	closeBlockTime = Sys_Microseconds();
 	
-	assert( logLevel > 0 );
+	//assert( logLevel > 0 );
```
Comment (kortemik) [inline neo/renderer/RenderLog.cpp]: This needs to be checked why it's disabled

Comment (damiel) [inline neo/renderer/RenderLog.cpp]: It was changed with commit 260756c957eea6cf267c93a6ae4c48efe44bf1a0

**Comment by kortemik** [comment]
----- R_InitOpenGL -----
Initializing OpenGL subsystem
Using 8 color bits, 24 depth, 8 stencil display
Using GLEW 1.10.0
Sys_InitInput: Joystick subsystem init
Sys_InitInput: Joystic - Found 0 joysticks
OpenGL Version   : 4.5
OpenGL Vendor    : NVIDIA Corporation
OpenGL Renderer  : Quadro 4000M/PCIe/SSE2
OpenGL GLSL      : 4.5
idCommon::VPrintf: truncated to 4094 characters
OpenGL Extensions: GL_AMD_multi_draw_indirect GL_ARB_arrays_of_arrays GL_ARB_base_instance GL_ARB_blend_func_extended GL_ARB_buffer_storage GL_ARB_clear_buffer_object GL_ARB_clear_texture GL_ARB_clip_control GL_ARB_color_buffer_float GL_ARB_compatibility GL_ARB_compressed_texture_pixel_storage GL_ARB_conservative_depth GL_ARB_compute_shader GL_ARB_compute_variable_group_size GL_ARB_conditional_render_inverted GL_ARB_copy_buffer GL_ARB_copy_image GL_ARB_cull_distance GL_ARB_debug_output GL_ARB_depth_buffer_float GL_ARB_depth_clamp GL_ARB_depth_texture GL_ARB_derivative_control GL_ARB_direct_state_access GL_ARB_draw_buffers GL_ARB_draw_buffers_blend GL_ARB_draw_indirect GL_ARB_draw_elements_base_vertex GL_ARB_draw_instanced GL_ARB_enhanced_layouts GL_ARB_ES2_compatibility GL_ARB_ES3_compatibility GL_ARB_ES3_1_compatibility GL_ARB_ES3_2_compatibility GL_ARB_explicit_attrib_location GL_ARB_explicit_uniform_location GL_ARB_fragment_coord_conventions GL_ARB_fragment_layer_viewport GL_ARB_fragment_program GL_ARB_fragment_program_shadow GL_ARB_fragment_shader GL_ARB_framebuffer_no_attachments GL_ARB_framebuffer_object GL_ARB_framebuffer_sRGB GL_ARB_geometry_shader4 GL_ARB_get_program_binary GL_ARB_get_texture_sub_image GL_ARB_gpu_shader5 GL_ARB_gpu_shader_fp64 GL_ARB_gpu_shader_int64 GL_ARB_half_float_pixel GL_ARB_half_float_vertex GL_ARB_imaging GL_ARB_indirect_parameters GL_ARB_instanced_arrays GL_ARB_internalformat_query GL_ARB_internalformat_query2 GL_ARB_invalidate_subdata GL_ARB_map_buffer_alignment GL_ARB_map_buffer_range GL_ARB_multi_bind GL_ARB_multi_draw_indirect GL_ARB_multisample GL_ARB_multitexture GL_ARB_occlusion_query GL_ARB_occlusion_query2 GL_ARB_parallel_shader_compile GL_ARB_pipeline_statistics_query GL_ARB_pixel_buffer_object GL_ARB_point_parameters GL_ARB_point_sprite GL_ARB_program_interface_query GL_ARB_provoking_vertex GL_ARB_query_buffer_object GL_ARB_robust_buffer_access_behavior GL_ARB_robustness GL_ARB_sample_shading GL_ARB_sampler_objects GL_ARB_seamless_cube_map GL_ARB_separate_shader_objects GL_ARB_shader_atomic_counter_ops GL_ARB_shader_atomic_counters GL_ARB_shader_bit_encoding GL_ARB_shader_draw_parameters GL_ARB_shader_group_vote GL_ARB_shader_image_load_store GL_ARB_shader_image_size GL_ARB_shader_objects GL_ARB_shader_precision GL_ARB_shader_storage_buffer_object GL_ARB_shader_subroutine GL_ARB_shader_texture_image_samples GL_ARB_shader_texture_lod GL_ARB_shading_language_100 GL_ARB_shading_language_420pack GL_ARB_shading_language_include GL_ARB_shading_language_packing GL_ARB_shadow GL_ARB_sparse_buffer GL_ARB_sparse_texture GL_ARB_stencil_texturing GL_ARB_sync GL_ARB_tessellation_shader GL_ARB_texture_barrier GL_ARB_texture_border_clamp GL_ARB_texture_buffer_object GL_ARB_texture_buffer_object_rgb32 GL_ARB_texture_buffer_range GL_ARB_texture_compression GL_ARB_texture_compression_bptc GL_ARB_texture_compression_rgtc GL_ARB_texture_cube_map GL_ARB_texture_cube_map_array GL_ARB_texture_env_add GL_ARB_texture_env_combine GL_ARB_texture_env_crossbar GL_ARB_texture_env_dot3 GL_ARB_texture_float GL_ARB_texture_gather GL_ARB_texture_mirror_clamp_to_edge GL_ARB_texture_mirrored_repeat GL_ARB_texture_multisample GL_ARB_texture_non_power_of_two GL_ARB_texture_query_levels GL_ARB_texture_query_lod GL_ARB_texture_rectangle GL_ARB_texture_rg GL_ARB_texture_rgb10_a2ui GL_ARB_texture_stencil8 GL_ARB_texture_storage GL_ARB_texture_storage_multisample GL_ARB_texture_swizzle GL_ARB_texture_view GL_ARB_timer_query GL_ARB_transform_feedback2 GL_ARB_transform_feedback3 GL_ARB_transform_feedback_instanced GL_ARB_transform_feedback_overflow_query GL_ARB_transpose_matrix GL_ARB_uniform_buffer_object GL_ARB_vertex_array_bgra GL_ARB_vertex_array_object GL_ARB_vertex_attrib_64bit GL_ARB_vertex_attrib_binding GL_ARB_vertex_buffer_object GL_ARB_vertex_program GL_ARB_vertex_shader GL_ARB_vertex_type_10f_11f_11f_rev GL_ARB_vertex_type_2_10_10_10_rev GL_ARB_viewport_array GL_ARB_window_pos GL_ATI_draw_buffers GL_ATI_texture_float GL_ATI_texture_mirror_once GL_S3_s3tc GL_EXT_texture_env_add GL_EXT_abgr GL_EXT_bgra GL_EX
   maxTextureAnisotropy: 16.000000
...using GL_EXT_texture_lod_bias
X..GL_GREMEDY_string_marker not found
...using GL_ARB_framebuffer_object
...using GL_EXT_framebuffer_blit
----- Initializing Render Shaders -----
WARNING: file renderprogs/tonemap.pixel, line 219: missing #endif
WARNING: file renderprogs/tonemap.pixel, line 219: missing #endif
WARNING: file renderprogs/tonemap.pixel, line 219: missing #endif
WARNING: file renderprogs/bink.vertex, line 192: missing closing brace
WARNING: file renderprogs/bink.vertex, line 71: invalid output parameter result.texcoord1
WARNING: file renderprogs/bink.vertex, line 72: invalid output parameter result.texcoord1
WARNING: file renderprogs/bink.vertex, line 73: invalid output parameter result.texcoord1
WARNING: file renderprogs/bink.vertex, line 86: invalid output parameter result.texcoord4
WARNING: file renderprogs/bink.vertex, line 87: invalid output parameter result.texcoord4
WARNING: file renderprogs/bink.vertex, line 88: invalid output parameter result.texcoord4
WARNING: file renderprogs/bink.vertex, line 91: invalid output parameter result.texcoord5
WARNING: file renderprogs/bink.vertex, line 92: invalid output parameter result.texcoord5
WARNING: file renderprogs/bink.vertex, line 93: invalid output parameter result.texcoord5
WARNING: file renderprogs/bink.vertex, line 107: invalid output parameter result.texcoord6
WARNING: file renderprogs/bink.vertex, line 108: invalid output parameter result.texcoord6
WARNING: file renderprogs/bink.vertex, line 109: invalid output parameter result.texcoord6
WARNING: file renderprogs/bink.vertex, line 110: invalid output parameter result.texcoord6
WARNING: file renderprogs/bink.vertex, line 115: invalid output parameter result.color
While compiling vertex program renderprogs/bink.vertex
-----------------
  1: // filename renderprogs/bink.vertex
  2: #version 150
  3: #define PC
  4: 
  5: float saturate( float v ) { return clamp( v, 0.0, 1.0 ); }
  6: vec2 saturate( vec2 v ) { return clamp( v, 0.0, 1.0 ); }
  7: vec3 saturate( vec3 v ) { return clamp( v, 0.0, 1.0 ); }
  8: vec4 saturate( vec4 v ) { return clamp( v, 0.0, 1.0 ); }
  9: vec4 tex2Dlod( sampler2D sampler, vec4 texcoord ) { return textureLod( sampler, texcoord.xy, texcoord.w ); }
 10: 
 11: 
 12: uniform vec4 _va_[13];
 13: 
 14: float dot3 (vec3 a , vec3 b ) {return dot ( a , b ) ; }
 15: float dot3 (vec3 a , vec4 b ) {return dot ( a , b. xyz ) ; }
 16: float dot3 (vec4 a , vec3 b ) {return dot ( a. xyz , b ) ; }
 17: float dot3 (vec4 a , vec4 b ) {return dot ( a. xyz , b. xyz ) ; }
 18: float dot4 (vec4 a , vec4 b ) {return dot ( a , b ) ; }
 19: float dot4 (vec2 a , vec4 b ) {return dot ( vec4 ( a , 0 , 1 ) , b ) ; }
 20: vec4 swizzleColor (vec4 c ) {
 21: 	return c ;
 22: }
 23: 
 24: in vec4 in_Position;
 25: in vec2 in_TexCoord;
 26: in vec4 in_Normal;
 27: in vec4 in_Tangent;
 28: in vec4 in_Color;
 29: 
 30: out vec2 vofi_TexCoord0;
 31: 
 32: void main() {
 33: 	gl_Position . x = dot4 ( in_Position , _va_[9 /* rpMVPmatrixX */] ) ;
 34: 	gl_Position . y = dot4 ( in_Position , _va_[10 /* rpMVPmatrixY */] ) ;
 35: 	gl_Position . z = dot4 ( in_Position , _va_[11 /* rpMVPmatrixZ */] ) ;
 36: 	gl_Position . w = dot4 ( in_Position , _va_[12 /* rpMVPmatrixW */] ) ;
 37: void main() {
 38: 		vec4 vNormal = in_Normal * 2.0 - 1.0 ;
 39: 		vec4 vTangent = in_Tangent * 2.0 - 1.0 ;
 40: 		vec3 vBitangent = cross ( vNormal. xyz , vTangent. xyz ) * vTangent. w ;
 41: 		vec4 modelPosition = in_Position ;
 42: 		vec3 normal = vNormal. xyz ;
 43: 		vec3 tangent = vTangent. xyz ;
 44: 		vec3 bitangent = vBitangent. xyz ;
 45: 		gl_Position . x = dot4 ( modelPosition , _va_[9 /* rpMVPmatrixX */] ) ;
 46: 		gl_Position . y = dot4 ( modelPosition , _va_[10 /* rpMVPmatrixY */] ) ;
 47: 		gl_Position . z = dot4 ( modelPosition , _va_[11 /* rpMVPmatrixZ */] ) ;
 48: 		gl_Position . w = dot4 ( modelPosition , _va_[12 /* rpMVPmatrixW */] ) ;
 49: 		vec4 defaultTexCoord = vec4 ( 0.0 , 0.5 , 0.0 , 1.0 ) ;
 50: 		vec4 toLight = normalize ( vec4 ( 0.0 , 0.5 , 1.0 , 1.0 ) ) ;
 51: 		vofi_TexCoord0 . x = dot3 ( tangent , toLight ) ;
 52: 		vofi_TexCoord0 . y = dot3 ( bitangent , toLight ) ;
 53: 		vofi_TexCoord0 . z = dot3 ( normal , toLight ) ;
 54: 		vofi_TexCoord0 . w = 1.0 ;
 55: 		result.texcoord1 = defaultTexCoord ;
 56: 		result.texcoord1 . x = dot4 ( in_TexCoord . xy , _va_[1 /* rpBumpMatrixS */] ) ;
 57: 		result.texcoord1 . y = dot4 ( in_TexCoord . xy , _va_[2 /* rpBumpMatrixT */] ) ;
 58: 		result.texcoord4 = defaultTexCoord ;
 59: 		result.texcoord4 . x = dot4 ( in_TexCoord . xy , _va_[3 /* rpDiffuseMatrixS */] ) ;
 60: 		result.texcoord4 . y = dot4 ( in_TexCoord . xy , _va_[4 /* rpDiffuseMatrixT */] ) ;
 61: 		result.texcoord5 = defaultTexCoord ;
 62: 		result.texcoord5 . x = dot4 ( in_TexCoord . xy , _va_[5 /* rpSpecularMatrixS */] ) ;
 63: 		result.texcoord5 . y = dot4 ( in_TexCoord . xy , _va_[6 /* rpSpecularMatrixT */] ) ;
 64: 		toLight = normalize ( toLight ) ;
 65: 		vec4 toView = normalize ( _va_[0 /* rpLocalViewOrigin */] - modelPosition ) ;
 66: 		vec4 halfAngleVector = toLight + toView ;
 67: 		result.texcoord6 . x = dot3 ( tangent , halfAngleVector ) ;
 68: 		result.texcoord6 . y = dot3 ( bitangent , halfAngleVector ) ;
 69: 		result.texcoord6 . z = dot3 ( normal , halfAngleVector ) ;
 70: 		result.texcoord6 . w = 1.0 ;
 71: 		result.color = ( swizzleColor ( in_Color ) * _va_[7 /* rpVertexColorModulate */] ) + _va_[8 /* rpVertexColorAdd */] ;
 72: 	}
-----------------
0(37) : error C0000: syntax error, unexpected '{', expecting ',' or ';' at token "{"
0(53) : error C1031: swizzle mask element not present in operand "z"
0(53) : warning C1068: array index out of bounds
0(54) : error C1031: swizzle mask element not present in operand "w"
0(54) : warning C1068: array index out of bounds

OpenTechEngine: /home/kordex/code/OpenTechBFG/neo/renderer/RenderProgs_GLSL.cpp:2219: void BFG::idRenderProgManager::LoadGLSLProgram(int, int, int): Assertion `prog.vertexUniformArray != -1 || vertexShaderIndex < 0 || vertexShaders[vertexShaderIndex].uniforms.Num() == 0' failed.
signal caught: Aborted
si_code -6
Trying to exit gracefully..

**Comment by damiel** [comment]
Needing to copy over from renderprogs should be fixed now, pls confirm.

**Comment by damiel** [comment]
Rebasing my render branch with master seems to break shadow(maps?). Ill have to investigate a bit why this is happening.

**Comment by damiel** [comment]
I am closing this pullrequest in my new and slimmed down one here: #130

**Pull request closed without merging**